src/libgrate/grate-texture.c

```diff
@@ -184,3 +264,212 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
 	if (err < 0)
 		grate_error("host1x_gr2d_clear() failed: %d\n", err);
 }
+
+static unsigned log2_size(unsigned size)
+{
+	switch (size) {
+	default:
+		return 11;
+	case 513 ... 1024:
+		return 10;
+	case 257 ... 512:
+		return 9;
+	case 129 ... 256:
+		return 8;
+	case 65 ... 128:
+		return 7;
+	case 33 ... 64:
+		return 6;
+	case 17 ... 32:
+		return 5;
+	case 9 ... 16:
+		return 4;
+	case 5 ... 8:
+		return 3;
+	case 3 ... 4:
+		return 2;
+	case 2:
+		return 1;
+	case 0 ... 1:
+		return 0;
+	}
```

Comment:
Yeah, one obvious optimization would be getting rid of CMA allocation on each command submit in the kernel driver. That gave a noticeable boost in performance while I was tackling the opentegra driver, actually not so long time ago. Also the current in-kernel command stream validation code completely kills performance, I usually have it turned off while hacking-developing.